plugins/metrics/src/main/java/org/apache/cloudstack/metrics/MetricsServiceImpl.java

```diff
@@ -335,61 +362,144 @@ public InfrastructureResponse listInfrastructure() {
+                hostMetrics.incrTotalResources();
+                updateHostMetrics(hostMetrics, hostJoinDao.findById(host.getId()));
             }
 
             metricsResponse.setState(clusterResponse.getAllocationState(), clusterResponse.getManagedState());
-            metricsResponse.setResources(metrics.getUpResources(), metrics.getTotalResources());
-            // CPU
-            metricsResponse.setCpuTotal(metrics.getTotalCpu());
-            metricsResponse.setCpuAllocated(metrics.getCpuAllocated(), metrics.getTotalCpu());
-            if (metrics.getCpuUsedPercentage() > 0L) {
-                metricsResponse.setCpuUsed(metrics.getCpuUsedPercentage(), metrics.getTotalHosts());
-                metricsResponse.setCpuMaxDeviation(metrics.getMaximumCpuUsage(), metrics.getCpuUsedPercentage(), metrics.getTotalHosts());
+            metricsResponse.setResources(hostMetrics.getUpResources(), hostMetrics.getTotalResources());
+            // add CPU and Memory metrics
+            addHostCpuMetricsToResponse(metricsResponse, clusterId, hostMetrics);
+            addHostMemoryMetricsToResponse(metricsResponse, clusterId, hostMetrics);
+
+            metricsResponse.setHasAnnotation(clusterResponse.hasAnnotation());
+            metricsResponses.add(metricsResponse);
+        }
+        return metricsResponses;
+    }
+
+    private void addHostMemoryMetricsToResponse(ClusterMetricsResponse metricsResponse, Long clusterId, HostMetrics hostMetrics) {
+        metricsResponse.setMemTotal(hostMetrics.getTotalMemory());
+        metricsResponse.setMemAllocated(hostMetrics.getMemoryAllocated(), hostMetrics.getTotalMemory());
+        if (hostMetrics.getMemoryUsed() > 0L) {
+            metricsResponse.setMemUsed(hostMetrics.getMemoryUsed(), hostMetrics.getTotalMemory());
+            metricsResponse.setMemMaxDeviation(hostMetrics.getMaximumMemoryUsage(), hostMetrics.getMemoryUsed(), hostMetrics.getTotalHosts());
+        }
+        // Memory thresholds
+        final Double memoryThreshold = AlertManager.MemoryCapacityThreshold.valueIn(clusterId);
+        final Float memoryDisableThreshold = DeploymentClusterPlanner.ClusterMemoryCapacityDisableThreshold.valueIn(clusterId);
+        metricsResponse.setMemoryUsageThreshold(hostMetrics.getMemoryUsed(), hostMetrics.getTotalMemory(), memoryThreshold);
+        metricsResponse.setMemoryUsageDisableThreshold(hostMetrics.getMemoryUsed(), hostMetrics.getTotalMemory(), memoryDisableThreshold);
+        metricsResponse.setMemoryAllocatedThreshold(hostMetrics.getMemoryAllocated(), hostMetrics.getTotalMemory(), memoryThreshold);
+        metricsResponse.setMemoryAllocatedDisableThreshold(hostMetrics.getMemoryAllocated(), hostMetrics.getTotalMemory(), memoryDisableThreshold);
+    }
+
+    private void addHostCpuMetricsToResponse(ClusterMetricsResponse metricsResponse, Long clusterId, HostMetrics hostMetrics) {
+        metricsResponse.setCpuTotal(hostMetrics.getTotalCpu());
```

Comment:
These `hostMetrics.get...` are called several times in this method, we could extract them to variables.